os/transaction: page align write data buffers to improve performance - version 2

bill-scales opened the pull request:
Align write data in Objectstore::Transaction encoded buffers so that RepOp and ECSubOpWrite messages align data on the receiving OSD to avoid a later memmove.  Also fix ECSubOpReadReply messages in a similar way so that read data is aligned on the receiving OSD.

Version 2 - A more complicated set of changes to Transaction.h to allow all write data to be aligned by avoiding placing the data buffer length in the data buffer. This requires the transaction to be created with two sets of bufferlists, one for the old format (used for backwards compatibility) and one for the new format.

## More detail

See https://github.com/ceph/ceph/pull/57738 (version 1) for details of why data alignment is a performance issue and the concept of encoding/decoding ObjectStore::Transaction's into the payload buffer and data buffer to allow data to be aligned.

Version 2 resolves the issue of version 1 by putting lengths in the payload buffer rather than the data buffer and by correcting the alignment of all writes in the transaction, not just the longest write. The challenge is how to do this while still being able to create backwards compatible encodings that include the lengths interspersed with write data. It solves this problem by creating two sets of buffer lists when an ObjectStore::Transaction is build - one set will be used to encode messages for downlevel OSDs (and has lengths interspersed with data), the other set will be used to encode messages for OSDs running new code (the reference counting in Ceph buffers means there are low overheads for having two sets of bufferlists). On the sending OSD the transaction maintains both formats, this is called a dual format transaction. On the receiving OSD the transaction will be either old or new format depending what encoding was used by the sending OSD. Code such as the iterator need to cope with all formats.

This code does not translate between old and new format (this is technically possible, but gets very messy because you have to iterate across the transaction and build a new set of bufferlists). Because of this there are two limitations on mixing different format transactions which will cause an assert:

### Limitation on append

It is not possible to append a new format and old format transaction or versa versa. Transactions of the same format or if either one of the transactions is in dual format can be appended. Code inspection shows the only current use of Transaction::append is for creating EC rollback transactions, both these transactions will be in dual format. There is also one existing test program that tests Transaction::append which also only uses transactions in dual format. In practice no code is going to receive transactions from two different OSDs (one running new code and one running old code) and then try and append these transactions together, so this limitations is not going to be an issue.

### Limitation on encode

It is not possible to encode a new format transaction in old format. Code inspection shows the only current uses of Transaction::encode/decode is when sending inter-OSD messages and by the ceph-dencoder tool. In practice no code is going to receive a transaction in a message from another OSD and then attempt to encode the transaction in old format to send to a downlevel OSD, so this limitation won’t be an issue.


Version 2 gets rid of yucky code in transaction.h to track the largest block of data in the Transaction, new code in write() optimizes the alignment of write data being added to the transaction and the iterator function decode_bl performs the reverse transformation. The extra code to deal with different transaction formats is not beautiful, but a nice feature of the implementation is that this is only required until backwards compatibility with older code levels is dropped.

New unit tests are added to fully exercise building, encoding, decoding, iterating and appending ObjectStore transactions and to validate the improvement in aligning buffers. The rbd bench utility is extended with extra options to provide more flexibility in generating I/O requests.

  The RepOp, ECSubWriteOp and ECSubReadReplyOp messages have been modified to align the write data buffers they send. Manual testing has confirmed that these buffers are correctly page aligned on the receiving OSD.

 The performance benchmark ceph_perf_objectstore was run on old and new code and shows negligible difference in performance for the typical set of operations performed on ObjectStore::Transaction’s.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

bill-scales commented on src/include/encoding.h:
```diff
@@ -117,7 +117,18 @@ inline void decode(bool &v, bufferlist::const_iterator& p) {
     ::ceph::decode_raw(e, p);						\
     v = e;								\
-  }
+  }									\
```
Reviewer feedback requested:

I've only implemented a minimal set of encode/decode templates using two buffers which is sufficient for ObjectStore::Transaction and the messages MOSDRepOp, MOSDECSubWrite and MOSDECSubReadReply. Should the addition of further templates be left until they are required or should this pull request cover the other standard types in encoding.h and denc.h?

rzarzynski replied:
Nah, I wouldn't add things until they are really necessary.
With this being said, I'm thinking on whether we truly need to (extensively) touch the dencoding framework.

bill-scales commented on src/messages/MOSDECSubOpReadReply.h:
```diff
@@ -63,7 +68,12 @@ class MOSDECSubOpReadReply : public MOSDFastDispatchOp {
     encode(pgid, payload);
     encode(map_epoch, payload);
-    encode(op, payload);
+    if (!HAVE_FEATURE(features, SERVER_SQUID)) {
```
Reviewer feedback requested:

The PR is currently using the feature SERVER_SQUID to select new/old version encoding for MOSDECSubOpReadReply, MOSDECSubOpWrite and MOSDRepOp. Should this be SERVER_TENTACLE? Is there a better way of checking compatibility of remote OSDs?

bill-scales commented on src/messages/MOSDECSubOpReadReply.h:
```diff
@@ -29,7 +29,7 @@ class MOSDECSubOpReadReply : public MOSDFastDispatchOp {
 
   int get_cost() const override {
-    return 0;
+    return data.length();
```
Reviewer feedback requested:

MOSDECSubOpReadReply and MOSDECSubOpWrite now encode write data into the data bl of the message, and now return the length of this buffer for get_cost(), previously these messages returned a cost of 0. This will change how these messages are prioritized by the scheduler. Should they return a cost of 0 instead to preserve old behavior, or was the old behavior a bug?

An aside is that all the ECSubOp* messages and the RepOp message unnecessarily override get_cost when the default method which returns data.length() would provide the same behavior - should these unnecessary overrides be deleted?

bill-scales commented on src/messages/MOSDRepOp.h:
```diff
@@ -188,6 +206,10 @@ class MOSDRepOp final : public MOSDFastDispatchOp {
     out << ")";
   }
+
+  void clear_buffers() override {
```
Reviewer feedback requested:

I’ve moved an ObjectStore::Transaction into the MOSDRepOp message so that it can be encoded when the message is sent rather than when the message is built. I added a clear_buffers() method (copying what is done in MOSDECSubOpWrite) to clean up the transaction and avoid buffers being held referenced. Is this correct?

bill-scales commented on src/os/Transaction.h:
```diff
@@ -248,13 +254,13 @@ class Transaction {
-    auto dp = nbl.cbegin();
-    decode(dp);
-  }
+  //  explicit Transaction(ceph::buffer::list::const_iterator &dp) {
```
Reviewer feedback requested:

I've deleted these helper functions to decode Transactions because there are now two decode functions (with one or two buffers) and I wanted code to be more explicit about which decode function to use. These functions were not being used by the current code.

Just want to check that there isn't a coding convention for having these type of decode helper functions?

bill-scales reviewed (commented):
I've annotated a few details where I'd appreciate some feedback form reviewers

idryomov reviewed (commented):
Same as in https://github.com/ceph/ceph/pull/57738, I'd suggest breaking additions to `rbd bench` command out into a separate PR for easier review.

bill-scales commented:
Here is a profile of a build containing the version 2 fix:

![Screenshot 2024-05-31 at 13 44 37](https://github.com/ceph/ceph/assets/156200352/ebd43384-e1b4-4089-95b1-9873259884e8)

and for reference the profile before the fix:

![Screenshot 2024-05-28 at 08 12 13](https://github.com/ceph/ceph/assets/156200352/32ad42a2-98ec-4006-9106-1bdc6bc32ae4)

We can see that the fix has had the desired effect of eliminating the memmove to realign data without causing any other adverse effects to performance when decoding and applying the transaction.

rzarzynski commented on src/include/encoding.h:
```diff
@@ -117,7 +117,18 @@ inline void decode(bool &v, bufferlist::const_iterator& p) {
     v = e;								\
-  }
+  }									\
+  inline void encode(type v, ::ceph::bufferlist& p_bl, ::ceph::bufferlist& d_bl,uint64_t features=0) { \
```
nit: space after comma.

rzarzynski commented on src/include/encoding.h:
```diff
@@ -276,6 +292,13 @@ inline void encode(const buffer::ptr& bp, bufferlist& bl)
+  __u32 len = bp.length();
+  encode(len, p_bl);
+  if (len)
+    d_bl.append(bp);
```
IIUC the main driving factor behind the new primitives is separating _head_ (typically size) from _non-head_ (actual data).
If that's the case, maybe we could reuse the `{encode, decode}_nohead` variants to save us some code duplication.

The obvious problem with applying this on broad scale is about containers and preserving the _head vs no-head_ distinction down the nested encodes. However, do we really need to worry? Isn't the optimization targeting basically several instances of bufferlist known to carry big data?

bill-scales replied:
The reason why I ended up creating new primitives is because of the ECSubReadReply structure/message. This needs to encode/decode a map that contains a list that contains pairs that have the data buffers that we want to keep aligned.

<pre>
  std::map<hobject_t, std::list<std::pair<uint64_t, ceph::buffer::list> >> buffers_read;
</pre>

We could write a custom encode/decode method for this structure, duplicating code from encoding.h/denc.h to handle the map, list, pair and buffer::list and call encode/decode_nohead for the data buffers we want to align, but this will end up duplicating code outside of the header files. My viewpoint is that it is better to keep all the encode/decode for containers in one place in the header files, rather than spread custom versions of this through the code, especially as I suspect we will find other places in the code base where this technique can be used to improve data buffer alignment.

rzarzynski replied:
I see your points. Have you considered an approach with `if constexpr()` and a _head-nohead trait_ (a template parameter; we already make use of the _traits_ concept in the dencoding frameworks) passed down the stack? It would require two passes (one with matadata bl, one with data bl) but maybe we could simplify even already existing code (by streamlining `{de,en}code_nohead()`).

Just thinking loudly.

bill-scales replied:
I really liked that idea and did a bit of prototyping. My observations:

1. The extra if statements in the {de,en}code functions are a bit messy, but perhaps 1 template with some ifs is still better than 2 templates for each type
2. I could write generic templates for encode_nohead, encode_headonly, encode that pass a different trait to the encode templates which means some of the nohead templates could be deleted
3. Decoding doesn't work :-( For compound types like a map you can't do the decode in two passes, you need to interleave the decode of the two buffers. This is also why decode_nohead requires extra parameters.

But the prototyping taught me that I don't need to duplicate all the templates - I can just have {de,en}code templates for types that take 2 bufferlists for payload and data buffers and then write a generic template like this to continue supporting encoding with a single bufferlist:

`template<class T> inline void encode(T& t,bufferlist &bl) { encode(t,bl,bl); }
`

This will tidy things up a lot and I think achieves what you were looking for. Will do this for the next commit

rzarzynski reviewed (commented):
Review-in-progress.

rzarzynski reviewed (commented):
In progress.

bill-scales commented on src/messages/MOSDECSubOpReadReply.h:
```diff
@@ -20,7 +20,7 @@
 private:
-  static constexpr int HEAD_VERSION = 2;
+  static constexpr int HEAD_VERSION = 3;
   static constexpr int COMPAT_VERSION = 1;
```
Good point, but we are ok:

* The decode function can decode all 3 versions.
* The encode function can encode version 2 (compatible with version 1 as it is just appending a field) or version 3. It uses a feature bit to determine when the receiving OSD can cope with version 3

I am however being lax about setting the compat_verison - this is always being set to 1. When sending a version 3 message compat_version should be set to 3 because this cannot be decoded by OSDs running version 1 or 2. We should never send them a version 3 message, but it would be better to set compat_version so they assert if this happened. I'll fix that

athanatos commented:
@rzarzynski @bill-scales Is this still in progress?

rzarzynski commented:
@athanatos, @bill-scales: IIRC the controversial / complex thing was `os::Transaction` started encoding ops in two formats the same time, just to tackle the upgrade case.

I hope https://github.com/ceph/ceph/commit/7858030a06212e160da5a3c635d54bfb5848556e and https://github.com/ceph/ceph/pull/59979 could help with that.

rzarzynski commented:
We talked off-line about this one. If my memory serves me right, @bill-scales is going to revive the PR squeezing the double encoding out.

bill-scales commented:
Working on cleaning this up.

1. Current version uses the peer OSD version when sending a transaction to another OSD to work out what format transaction to encode. This gets messy when OSDs are running different code levels - in particular for replica pools where ideally we want to encode the transaction once and send it to multiple OSDs. Next attempt will use min_peer_features which is a set features supported by all the OSDs in a PG and is calculated early in Peering (before I/O starts) to decide how to encode the transactions

2. Current version makes quite a few changes to encoding.h to allow data structures such as sets of buffers to be encoded with buffers encoded into a data buffer and the rest of the data encoded into a metadata buffer. This turns out to be a lot of boilerplate code and still requires changes when encoding a data structure because it isn't really possible to distinguish data buffers for things like attributes from data buffers for I/O. The code will be simpler just by writing a special case encode function for the ECSubReadReply message.

rzarzynski commented:
@bill-scales: I pushed the simplified encoding of `os::Transaction` based on https://github.com/ceph/ceph/commit/7858030a06212e160da5a3c635d54bfb5848556e. Now we encode in single format only. Would you mind reviewing?

bill-scales commented:
New version of the code addressing review issues:

1. Removed rbd-bench changes - moved to separate PR
2. Undone changes to cost in transaction - will raise tracker issue to look into this
3. Removed all the changes to encoding.h. This means EC sub read reply has to manually encode/decode a map of list of result buffers to separate out payload and data, but overall it simplifies the change.
4. Neatened up encoding for messages by passing features into the encode function of the data structure and having this select the version and whether to separate payload and data.
5. Reverted changes to how MOSDRepOp is encoded. Previous versions passed the transaction to MOSDRepOp and deferred encoding the transaction until the message was about to be sent. This meant when replicating to multiple OSDs the transaction got encoded multiple times. Encoding the transaction just once is more efficient but means the MOSDRepOp message now has payload (MOSDRepOp payload), middle (os:Transaction payload) and data (os:Transaction data)

bill-scales commented:
jenkins test make check

rzarzynski reviewed (approved):
LGTM except a problem with `decode_nohead()` taking `length` as `int`. I'm going to fix it in the PR.

bill-scales reviewed (commented):
Reviewed Radik's change to decode_nohead and checked there are no new compiler warnings regarding sign comparisons because of the change. lgtm

bill-scales commented:
WARNING: Transaction.h has a #ifdef WITH_SEASTORE that will need to be updated to WITH_CRIMSON before being merged because of https://github.com/ceph/ceph/commit/23c33f69ff977f7a05d3e3368e078b20e67a5ced

bill-scales commented:
Rebased - only other change is s/WITH_SEASTAR/WITH_CRIMSON because of https://github.com/ceph/ceph/commit/23c33f69ff977f7a05d3e3368e078b20e67a5ced

bill-scales commented:
jenkins test make check arm64

bill-scales commented:
jenkins test make check arm64

ljflores commented on src/osd/ReplicatedBackend.cc:
```diff
@@ -1183,8 +1190,9 @@ void ReplicatedBackend::do_repop(OpRequestRef op)
 
-  auto p = const_cast<bufferlist&>(m->get_data()).cbegin();
-  decode(rm->opt, p);
+  auto p = const_cast<bufferlist&>(m->get_middle()).cbegin();
+  auto d = const_cast<bufferlist&>(m->get_data()).cbegin();
+  rm->opt.decode(m->get_middle().length() != 0 ?  p : d, d);
```
Seeing a teuthology error around this area of the code:
/a/lflores-2025-04-10_15:17:40-rados-wip-lflores-testing-2-2025-04-09-1400-distro-default-smithi/8234757
```
2025-04-10T18:16:11.557 INFO:tasks.ceph.osd.15.smithi135.stderr:terminate called after throwing an instance of 'ceph::buffer::v15_2_0::malformed_input'
2025-04-10T18:16:11.638 INFO:tasks.ceph.osd.15.smithi135.stderr:  what():  void ceph::os::Transaction::decode(ceph::buffer::v15_2_0::list::const_iterator&, ceph::buffer::v15_2_0::list::const_iterator&) decode past end of struct encoding: Malformed input [buffer:3]
2025-04-10T18:16:11.638 INFO:tasks.ceph.osd.15.smithi135.stderr:*** Caught signal (Aborted) **
2025-04-10T18:16:11.638 INFO:tasks.ceph.osd.15.smithi135.stderr: in thread 7fdbeb491640 thread_name:tp_osd_tp
2025-04-10T18:16:11.644 INFO:tasks.ceph.osd.15.smithi135.stderr: ceph version 20.0.0-1326-ga86028d6 (a86028d68e7921896709e151a2ea1a7c9fc0be14) tentacle (dev)
2025-04-10T18:16:11.644 INFO:tasks.ceph.osd.15.smithi135.stderr: 1: /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7fdc065b7520]
2025-04-10T18:16:11.644 INFO:tasks.ceph.osd.15.smithi135.stderr: 2: pthread_kill()
2025-04-10T18:16:11.644 INFO:tasks.ceph.osd.15.smithi135.stderr: 3: raise()
2025-04-10T18:16:11.644 INFO:tasks.ceph.osd.15.smithi135.stderr: 4: abort()
2025-04-10T18:16:11.644 INFO:tasks.ceph.osd.15.smithi135.stderr: 5: /lib/x86_64-linux-gnu/libstdc++.so.6(+0xa2bbe) [0x7fdc06948bbe]
2025-04-10T18:16:11.644 INFO:tasks.ceph.osd.15.smithi135.stderr: 6: /lib/x86_64-linux-gnu/libstdc++.so.6(+0xae24c) [0x7fdc0695424c]
2025-04-10T18:16:11.644 INFO:tasks.ceph.osd.15.smithi135.stderr: 7: /lib/x86_64-linux-gnu/libstdc++.so.6(+0xae2b7) [0x7fdc069542b7]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 8: /lib/x86_64-linux-gnu/libstdc++.so.6(+0xae518) [0x7fdc06954518]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 9: ceph-osd(+0x3c7b27) [0x55e8bcab1b27]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 10: (ReplicatedBackend::do_repop(boost::intrusive_ptr<OpRequest>)+0x477) [0x55e8bce2e537]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 11: (ReplicatedBackend::_handle_message(boost::intrusive_ptr<OpRequest>)+0x1d7) [0x55e8bce3c1d7]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 12: (PGBackend::handle_message(boost::intrusive_ptr<OpRequest>)+0x56) [0x55e8bce495f6]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 13: (PrimaryLogPG::do_request(boost::intrusive_ptr<OpRequest>&, ThreadPool::TPHandle&)+0x7ed) [0x55e8bcd963dd]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 14: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x197) [0x55e8bccd1837]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 15: (ceph::osd::scheduler::PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x69) [0x55e8bcf22339]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 16: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0xa76) [0x55e8bcce5b36]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 17: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x292) [0x55e8bd2412e2]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 18: ceph-osd(+0xb57864) [0x55e8bd241864]
2025-04-10T18:16:11.645 INFO:tasks.ceph.osd.15.smithi135.stderr: 19: /lib/x86_64-linux-gnu/libc.so.6(+0x94b43) [0x7fdc06609b43]
2025-04-10T18:16:11.646 INFO:tasks.ceph.osd.15.smithi135.stderr: 20: /lib/x86_64-linux-gnu/libc.so.6(+0x126a00) [0x7fdc0669ba00]
```

ljflores replied:
Testing ref: https://tracker.ceph.com/issues/70862

rzarzynski replied:
@bill-scales.

bill-scales commented:
Reproduced, debugged and fixed the issue causing QA failures.

Change is to MOSDop::encode_payload

```
- set_middle(txn_payload);
+ bufferlist middle(txn_payload);
+ set_middle(middle);
```

It was only teuthology tests that had messenger error inject that were failing. The error inject caused messenger to have two attempts at encoding and sending the message. The call to set_middle does a bufferlist assignment that empties the source bufferlist, so encoding the message twice meant the middle buffer wasn't sent causing a decode problem on the receiving osd. The fix makes a shallow copy of the bufferlist (will be a single buffer) so that it is safe to call encode_payload multiple times.

rzarzynski commented:
@ljflores: let's retake this into the T testing.

kamoltat commented:
RADOS APPROVED: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues70862

ljflores commented:
Crimson approved by @athanatos.

Failure, unrelated, tracked by:
- https://tracker.ceph.com/issues/70501
